### PR TITLE
Add tests regarding navigation inside sandboxed iframes.

### PR DESCRIPTION
**Pull request by fred-wang**
This commit adds basic tests for frame navigation to verify some cases handled by Chromium, that I'm trying to implement in WebKit [1] [2]. cc'ing @mikewest @RByers @japhet who have worked on this. Also cc'ing @cdumez and @bzbarsky for WebKit & Mozilla.

[1] https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/frame/LocalFrame.cpp?type=cs&q=LocalFrame::CanNavigateWithoutFramebusting
[2] https://bugs.webkit.org/show_bug.cgi?id=173162

### CODE REVIEW

**Comment by fred-wang** [comment]
cc'ing @natechapin too.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242347039)

# Firefox (nightly) #
Testing web-platform-tests at revision d17aa6b41cf3f10b8e29548c29885ef525157e97
Using browser at version BuildID 20170612100241; SourceStamp 27cad9749cddf68e11fdd4e5d73dad84a8f8cf23
Starting 10 test iterations
## Unstable results ##


|                                              Test                                              | Subtest |           Results           | Messages |
|------------------------------------------------------------------------------------------------|---------|-----------------------------|----------|
| `/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html` |         | **ERROR: 5/10, PASS: 5/10** |          |

## All results ##
<details>
<summary>5 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html</a></summary>


| Subtest |           Results           | Messages |
|---------|-----------------------------|----------|
|         | **ERROR: 5/10, PASS: 5/10** |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-1.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-1.html</a></summary>


|                            Subtest                             | Results | Messages |
|----------------------------------------------------------------|---------|----------|
|                                                                | OK      |          |
| `Check that sandboxed iframe can not navigate their ancestors` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-2.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-2.html</a></summary>


|                           Subtest                            | Results | Messages |
|--------------------------------------------------------------|---------|----------|
|                                                              | OK      |          |
| `Check that unsandboxed iframe can navigate their ancestors` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_descendants.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_descendants.html</a></summary>


|                           Subtest                            | Results | Messages |
|--------------------------------------------------------------|---------|----------|
|                                                              | OK      |          |
| `Check that sandboxed iframe can navigate their descendants` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_itself.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_itself.html</a></summary>


|                      Subtest                      | Results | Messages |
|---------------------------------------------------|---------|----------|
|                                                   | OK      |          |
| `Check that sandboxed iframe can navigate itself` | PASS    |          |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242347040)

# Sauce (safari) #
Testing web-platform-tests at revision d17aa6b41cf3f10b8e29548c29885ef525157e97
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>5 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-1.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-1.html</a></summary>


|                            Subtest                             | Results | Messages |
|----------------------------------------------------------------|---------|----------|
|                                                                | TIMEOUT |          |
| `Check that sandboxed iframe can not navigate their ancestors` | NOTRUN  |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-2.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-2.html</a></summary>


|                           Subtest                            | Results | Messages |
|--------------------------------------------------------------|---------|----------|
|                                                              | OK      |          |
| `Check that unsandboxed iframe can navigate their ancestors` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_descendants.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_descendants.html</a></summary>


|                           Subtest                            | Results | Messages |
|--------------------------------------------------------------|---------|----------|
|                                                              | OK      |          |
| `Check that sandboxed iframe can navigate their descendants` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_itself.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_itself.html</a></summary>


|                      Subtest                      | Results | Messages |
|---------------------------------------------------|---------|----------|
|                                                   | OK      |          |
| `Check that sandboxed iframe can navigate itself` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242347046)

# Chrome (unstable) #
Testing web-platform-tests at revision d17aa6b41cf3f10b8e29548c29885ef525157e97
Using browser at version 61.0.3124.4 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>5 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-1.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-1.html</a></summary>


|                            Subtest                             | Results | Messages |
|----------------------------------------------------------------|---------|----------|
|                                                                | OK      |          |
| `Check that sandboxed iframe can not navigate their ancestors` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-2.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-2.html</a></summary>


|                           Subtest                            | Results | Messages |
|--------------------------------------------------------------|---------|----------|
|                                                              | OK      |          |
| `Check that unsandboxed iframe can navigate their ancestors` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_descendants.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_descendants.html</a></summary>


|                           Subtest                            | Results | Messages |
|--------------------------------------------------------------|---------|----------|
|                                                              | OK      |          |
| `Check that sandboxed iframe can navigate their descendants` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_itself.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_itself.html</a></summary>


|                      Subtest                      | Results | Messages |
|---------------------------------------------------|---------|----------|
|                                                   | OK      |          |
| `Check that sandboxed iframe can navigate itself` | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6221/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html">/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | TIMEOUT |          |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/242347047)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision d17aa6b41cf3f10b8e29548c29885ef525157e97
Using browser at version 14.14393
Starting 10 test iterations

**Comment by fred-wang** [comment]
Note: I modified the test for allow-top-navigation to make it more reliable. If needed, we can add a similar test to check that top navigation is blocked when allow-top-navigation is unset, however this is already done by iframe_sandbox_allow_top_navigation_by_user_activation_without_user_gesture.html

File: html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    <script>
+      if (opener) {
+        // We're the popup (i.e. a top frame).  Load into the iframe the page
+        // trying to modifying the top frame and transmit the result to our
+        // opener.
+        onmessage = function(e) {
+          opener.postMessage({data: e.data, origin: e.origin}, "*")
+          window.close();
+        }
+        document.querySelector("iframe").src = "support/iframe-that-performs-top-navigation-without-user-gesture-passed.html";
```
Comment (mikewest) [inline html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html]: It seems like the timing here's not going to work. The frame loads `iframe-that-performs-top-navigation-without-user-gesture-passed.html`, which navigates the page containing the frame. Depending on when that navigation fires, there might or might not actually be a message handler to `postMessage()` up to the `opener`.

File: html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  </head>
+  <body>
+    <iframe sandbox="allow-top-navigation allow-scripts"></iframe>
+    <script>
+      if (opener) {
+        // We're the popup (i.e. a top frame).  Load into the iframe the page
+        // trying to modifying the top frame and transmit the result to our
+        // opener.
+        onmessage = function(e) {
+          opener.postMessage({data: e.data, origin: e.origin}, "*")
```
Comment (mikewest) [inline html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html]: Why pass the origin if you're not using it in the check on line 27?

File: html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        document.querySelector("iframe").src = "support/iframe-that-performs-top-navigation-without-user-gesture-passed.html";
+      } else {
+        // We are the main test page.  Open ourselves as a popup, so that we can
+        // can experiment navigation of the top frame.
+        var t = async_test();
+        onmessage = t.step_func_done(function(e) {
+          assert_equals(e.data.data, "PASS", "Should have the right message");
+        });
+        var popupWin = window.open();
+        popupWin.location.href = location.href;
```
Comment (mikewest) [inline html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation.html]: Tiny style nit: I'd write this as:

```
async_test(t => {
  window.addEventListener('message', t.step_func_done(e => {
    assert_equals(e.data.data, "PASS");
    e.source.close();
  }));
  window.open(location.href);
}, "Frames with `allow-top-navigation` should be able to navigate the top frame.");
```

Closing the window helps clean things up when you're running the test manually, and I don't think you need the extra `t` or `popupWin` variables.

**Comment by fred-wang** [comment]
@mikewest Thanks. I pushed another commit that hopefully fixes your concern. I finally also added a similar test when allow-top-navigation is unset.

**Comment by fred-wang** [comment]
@mikewest ¿review ping?

**Review by mikewest — APPROVED** [review]
LGTM % two small style nits.

**Comment by fred-wang** [comment]
@mikewest thanks!

**Pull request merged**